src/jit/codegenlegacy.cpp

```diff
@@ -19462,14 +19462,30 @@ regMaskTP CodeGen::genCodeForCall(GenTreePtr call, bool valUsed)
                         regNumber indCallReg;
 
                         case IAT_VALUE:
+                        {
                             //------------------------------------------------------
                             // Non-virtual direct calls to known addressess
                             //
                             // The vast majority of calls end up here....  Wouldn't
                             // it be nice if they all did!
                             CLANG_FORMAT_COMMENT_ANCHOR;
 #ifdef _TARGET_ARM_
-                            if (!arm_Valid_Imm_For_BL((ssize_t)addr))
+                            // We may use direct call for some of recursive calls
+                            // as we can safely estimate the distance from the call site to the top of the method
+                            const int MAX_PROLOG_SIZE = MAX_USHORT_SIZE_T;
+                            const int codeOffset      = getEmitter()->emitCurCodeOffset + getEmitter()->emitCurIGsize;
+                            if (compiler->gtIsRecursiveCall(call->AsCall()) &&
+                                MAX_PROLOG_SIZE + codeOffset + 4 < 16777216) // if code offset fits in imm24
```

Comment:
Why is the constant 4 added? That should be explained.
Also, in unwind.h we have `MAX_PROLOG_SIZE_BYTES = 40`. I wouldn't think we should be using different sizes for this.
Also, instead of using a literal constant for the maximum value, I would suggest adding a constant such as `CALL_DIST_MAX_NEG` in target.h (see where, e.g. LBL_DIST_SMALL_MAX_NEG etc. are declared).